Add a KubeVirt CR option to use a dedicated migration network

jean-edouard opened the pull request:
**What this PR does / why we need it**:
This PR adds the `dedicatedMigrationNetwork` migration option, to specify a multus network to use for migrations.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This requires a multus network already configured to link all the nodes together and provide a DHCP server.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Migrations can now be done over a dedicated multus network
```

rmohr commented on pkg/virt-handler/vm.go:
```diff
@@ -687,6 +687,46 @@ func (d *VirtualMachineController) migrationSourceUpdateVMIStatus(origVMI *v1.Vi
+		}
+		for _, addr := range addrs {
+			var ip net.IP
+			if iface.Name != "net1" {
```
Would it make sense to use the pod [downward api](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#capabilities-of-the-downward-api) to get `k8s.v1.cni.cncf.io/networks-status` annotation info into virt-handler and in addition a commandline flag for virt-handler where you pass in the multus network name?

One can then look up info like this:

```
    k8s.v1.cni.cncf.io/networks-status: |- 
      [{
          "name": "openshift-sdn",
          "interface": "eth0",
          "ips": [
              "10.128.2.14"
          ],
          "default": true,
          "dns": {}
      },{
          "name": "macvlan-bridge",
          "interface": "net1",
          "ips": [
              "20.2.2.100"
          ],
          "mac": "22:2f:60:a5:f8:00",
          "dns": {}
      }]
```

This would probably remove all guessing and you could simply read the IP from this info, since you would know the multus network name from the commandline flag.

jean-edouard replied:
Force-pushed switch to downward API, let me know what you think!
Haven't done the commandline part yet. Would it make sense to parse the json part of the NetworkAttachmentDefinition to find the network name and use that as a key instead of the interface name?

rmohr replied:
Yes, I think so. Just using the multus name as key and then extracting all info from there sounds great.

jean-edouard replied:
So, I started working on that, using https://github.com/k8snetworkplumbingwg/network-attachment-definition-client to access the NetworkAttachmentDefinition.
However, the client above fails to compile. I'm getting obscure errors I wouldn't know how to fix. (Starts with: `vendor/github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned/typed/k8s.cni.cncf.io/v1/k8s.cni.cncf.io_client.go:74:32: undefined: serializer.DirectCodecFactory`).
Even if it worked, finding the NetworkAttachmentDefinition and parsing the json blob it contains adds a lot of complxity to the code, so I'm not sure it's worth it...
Maybe we could request that the name of the network matches the name of the NetworkAttachmentDefinition? Or add fields in the CR? (we'll need a second one anyway if we want to allow using NetworkAttachmentDefinitions that are in a different namespace than `kubevirt`)
@rmohr please let me know what you think! Thanks

rmohr replied:
@jean-edouard did you consider simply parsing the json into a `map[string]interface{}`? But in general you don't need the client. You only need the api definition to deserialize the json. Consider simply copying the `NetworkStatus` into our codebase (https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/master/pkg/apis/k8s.cni.cncf.io/v1/types.go#L101). It is nice if we can reuse external code, but at the end there is a `v1` REST API defined by it, which means that we can define our own struct for the API without risks.

rmohr replied:
But looking at your latest code, you basically already did that. So maybe I did not understand what you try to do? I think in general you will not have to ever look up a `NetworkAttachmentDefinition` in the cluster if I did not miss anything.

jean-edouard replied:
This is probably a lack of knowledge on my side, but is there a way to retrieve a NetworkAttachmentDefinition from k8s in golang without having to use a client like network-attachment-definition-client?
If so, could you please point me to some docs/example on how to do that?
Thanks!

jean-edouard replied:
Ah, so maybe this is a misunderstanding. In the CR, the cluster admin provides a NetworkAttachmentDefinition by name.
Inside that object there is a json blob (don't ask me why), with a different name, and that's the name I need to parse the yaml network info!

jean-edouard replied:
This was discussed offline, and Roman may have found a specific notation that could potentially alleviate this hole issue.
I'm looking into that now.

jean-edouard replied:
Indeed, the network annotation can name the interface using the "@" notation. Pushed, thanks!

jean-edouard commented on pkg/virt-handler/vm.go:
```diff
@@ -713,20 +753,21 @@ func (d *VirtualMachineController) migrationTargetUpdateVMIStatus(vmi *v1.Virtua
 		hostAddress := ""
-
+		migrationIpAddress := d.ipAddress
+		findMultusIP(vmi, &migrationIpAddress)
```
Not sure what you mean there.. The main pod IP address is actually just passed as a commandline argument.

rmohr replied:
Ah I see, I misinterpreted what this does. Will add more precise suggestions.

rmohr commented on pkg/virt-handler/vm.go:
```diff
@@ -687,6 +690,31 @@ func (d *VirtualMachineController) migrationSourceUpdateVMIStatus(origVMI *v1.Vi
+}
+
+// Using the downward API, look for dedicated migration network net1 and, if found, set migration IP to its IP
+func findMigrationIP(vmi *v1.VirtualMachineInstance, migrationIp *string) {
```
Can you move this whole `findMigrationIP` to the virt-handler startup phase and just set d.ipAddress upfront on the controller? It can not change during the handler lifetime and it shoud keep the core-logic more in the focus then.

rmohr commented on pkg/virt-handler/vm.go:
```diff
@@ -687,6 +690,31 @@ func (d *VirtualMachineController) migrationSourceUpdateVMIStatus(origVMI *v1.Vi
+
+	dat, err := ioutil.ReadFile("/etc/podinfo/network-status")
+	if err != nil {
+		return
```
Let's see that we return errors. I would suggest that you detect that we are supposed to use a multus network during the startup phase and only then call this function. Then if you see an error, fail the whole handler start.

rmohr commented on pkg/virt-handler/vm.go:
```diff
@@ -687,6 +690,31 @@ func (d *VirtualMachineController) migrationSourceUpdateVMIStatus(origVMI *v1.Vi
+		return
+	}
+	for _, ns := range networkStatus {
+		if ns.Interface == "net1" && len(ns.Ips) > 0 {
```
As suggested by you in another comment, here it would be great if you would use the multus network name and extract all necessary info from there. Only this way it does not matter if we have one or more multus netwoks. Otherwise any kind of preset mechanism which would add multus annotations could cause trouble.

jean-edouard commented:
Depends on https://github.com/kubevirt/kubevirtci/pull/671

jean-edouard commented:
/retest

jean-edouard commented:
/hold
There's a couple things wrong with this PR, no need to waste resources testing/reviewing it, will unhold when fixed.

jean-edouard commented:
/hold cancel
- Functest moved to network module to get CNAO and whereabouts
- Last PR comments addressed in new commit
- Rebased

jean-edouard commented:
/hold

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

jean-edouard commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-network

jean-edouard commented:
/hold cancel
Assumed all along that sig-network requested at least 1 secondary network... Added.
Also moved to more harmless IP range. Sorry for all the noise!

jean-edouard commented:
/retest

jean-edouard commented:
/retest

jean-edouard commented:
/retest

jean-edouard commented:
/retest

rmohr commented on pkg/virt-config/config-map.go:
```diff
@@ -311,6 +311,7 @@ type migrationConfiguration struct {
 	UnsafeMigrationOverride           *bool              `json:"unsafeMigrationOverride,string,omitempty"`
 	AllowPostCopy                     *bool              `json:"allowPostCopy,string,omitempty"`
 	DisableTLS                        *bool              `json:"disableTLS,omitempty"`
+	DedicatedMigrationNetwork         *string            `json:"dedicatedMigrationNetwork,omitempty"`
```
I think you can remove this. We don't add new options to the config-map anymore. If people would try to set it, we would just ignore it and stick with the default.

jean-edouard replied:
I wish! The project won't compile if this struct is not identical to the CR struct.

rmohr replied:
Ack. With https://github.com/kubevirt/kubevirt/pull/6784 it will not have a relevance anymore.

rmohr commented on pkg/virt-handler/vm.go:
```diff
@@ -689,6 +692,33 @@ func (d *VirtualMachineController) migrationSourceUpdateVMIStatus(origVMI *v1.Vi
+}
+
+// Using the downward API, look for dedicated migration network migration0 and, if found, set migration IP to its IP
+func findMigrationIP(migrationIp *string) error {
```
I think we were going forth and back a little bit here on using the downwards api or directly reading the IP from the interface. What was the final reason for using the downward api (which is fine by me in general)?

jean-edouard replied:
I'm pretty sure that was an offline suggestion from you, and it does look cleaner using the downwards API. Not sure if there's much more to it.

rmohr replied:
Ack.

rmohr commented on pkg/virt-handler/vm.go:
```diff
@@ -1432,6 +1461,12 @@ func (c *VirtualMachineController) Run(threadiness int, stopCh chan struct{}) {
 	defer c.Queue.ShutDown()
 	log.Log.Info("Starting virt-handler controller.")
 
+	err := findMigrationIP(&c.migrationIpAddress)
```
Would it make sense to look the ip address up outside of the controller and pass it as the desired IP address instead of implicitly overriding the passed in IP address ad creation time? The rename to `migrationIPAddress` is great if that is really the only reason why we ever use it.

jean-edouard replied:
Done!

rmohr reviewed (commented):
A first pass over the code.

rmohr commented on automation/test.sh:
```diff
@@ -43,6 +43,7 @@ elif [[ $TARGET =~ sig-network ]]; then
   export KUBEVIRT_PROVIDER=${TARGET/-sig-network/}
   export KUBEVIRT_DEPLOY_ISTIO=true
   export KUBEVIRT_DEPLOY_CDI=false
+  export KUBEVIRT_NUM_SECONDARY_NICS=1
```
This is the default anyway, right?

jean-edouard replied:
No! I was surprised too

rmohr commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -323,11 +347,18 @@ func (app *virtHandlerApp) Run() {
 	}
 
+	migrationIpAddress := app.PodIpAddress
+	err = findMigrationIP(&migrationIpAddress)
```
Maybe let's avoid the pointer magic. How about something like this:


```
findMigrationIP(defaultMigrationIP string) (string, error) {}
```

and then 


```
migrationIPAddress, err := findMigrationIP(app.PodIPAddress)
```

rmohr commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -176,6 +179,27 @@ func (app *virtHandlerApp) markNodeAsUnschedulable(logger *log.FilteredLogger) {
+	if err != nil {
+		return fmt.Errorf("failed to read network status from downwards API")
+	}
+	err = yaml.Unmarshal(dat, &networkStatus)
```
Do we rely on some implicit behaviour that we can deal with empty files? Because it looks like we would crate this file always and in case that there is no annotation the file is presumably empty?

jean-edouard replied:
The file will always contain at least 1 entry for the main network.
Also yes, `yaml.Unmarshall()` will happily decode an empty string into an empty list.

rmohr replied:
Even if multus is not installed? The annotation looks very multus specific.

jean-edouard replied:
The annotation is multus-specific, but the network status isn't.
If the annotation is not present and/or multus is not installed, the network-status file will still show the status of the main network.
Or maybe I misunderstood your question? :)

rmohr replied:
Well maybe I misunderstand something, but 

```
	container.VolumeMounts = append(container.VolumeMounts, corev1.VolumeMount{
		Name:      "podinfo",
		MountPath: "/etc/podinfo",
	})
	pod.Volumes = append(pod.Volumes, corev1.Volume{
		Name: "podinfo",
		VolumeSource: corev1.VolumeSource{
			DownwardAPI: &corev1.DownwardAPIVolumeSource{
				Items: []corev1.DownwardAPIVolumeFile{
					{
						Path: "network-status",
						FieldRef: &corev1.ObjectFieldSelector{
							FieldPath: `metadata.annotations['k8s.v1.cni.cncf.io/network-status']`,
						},
					},
				},
			},
		},
	})
	```
	
seems to be the source of the `network-status` file and that one looks to me like it would only contain any context if this annotation exists. And this annotation does probably not exist if there is not multus installed?

jean-edouard replied:
Sorry, you are absolutely right!
Without multus, /etc/podinfo/network-status still gets created but is empty.
As I said before though, this case is properly handled by the code, though not explicitly.

jean-edouard replied:
Added unit tests

rmohr replied:
Thanks! :)

rmohr commented on cmd/virt-handler/virt-handler_test.go:
```diff
@@ -0,0 +1,90 @@
+			Expect(err.Error()).To(ContainSubstring("failed to read network status from downwards API"))
+		})
+		It("Should handle the empty file case", func() {
+			file, err := ioutil.TempFile("", "test")
```
I think we should clean up the temporary files. They pile up in the tmp directory otherwise.

jean-edouard replied:
There's a deferred removal below, do we need something else?

rmohr replied:
great then :)

jean-edouard commented:
> @jean-edouard @rmohr what needs to happen to converge this?

Sorry this is taking so long...
I just replied to a comment that I previously missed.
The PR also needs a rebase which I'll take care of right now.

rmohr commented:
/retest
/approve

@xpivarc do you want to review the PR a last time and lgtm?

xpivarc commented on tests/network/vmi_networking.go:
```diff
@@ -940,6 +940,70 @@ var _ = SIGDescribe("[Serial][rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com]
+				By("Setting it as the migration network in the KubeVirt CR")
+				tests.SetDedicatedMigrationNetwork(nad.Name)
+
+				// FIXME?: By design, changing migration settings trigger a re-creation of the virt-handler pods, amongst other things.
```
I think it could be done through conditions on the `kv` but I am also fine with your approach. I would just ask that we move the `Eventually` block into `SetDedicatedMigrationNetwork` if we ever going to use it again.

xpivarc commented on pkg/virt-operator/resource/generate/components/deployments.go:
```diff
@@ -138,6 +138,14 @@ func newPodTemplateSpec(podName string, imageName string, repository string, ver
 		},
 	}
 
+	if migrationNetwork != nil {
```
I would move this to `NewHandlerDaemonSet` as it is only needed there.

xpivarc reviewed (approved):
Great work! 
My only real concern is the `AfterEach` block.
/hold

jean-edouard commented:
Thank you @xpivarc !
I agree with all your comments and have addressed them in a new commit.
PTAL!

xpivarc commented on tests/migration.go:
```diff
@@ -80,34 +81,52 @@ func ConfirmVMIPostMigration(virtClient kubecli.KubevirtClient, vmi *v1.VirtualM
-	Expect(err).ToNot(HaveOccurred())
+	// By design, changing migration settings trigger a re-creation of the virt-handler pods, amongst other things.
+	//   However, even if SetDedicatedMigrationNetwork() calls UpdateKubeVirtConfigValueAndWait(), VMIs still seem to get scheduled on outdated virt-handler pods.
+	//   Not sure if it's a bug or a feature though. Waiting for all "old" virt-handlers to disappear ensures test VMIs will be created on updated virt-handler pods.
```
This is not a bug. `UpdateKubeVirtConfigValueAndWait` is only waiting for the new config to be propagated but that doesn't guarantee that operator had time to do the job.

xpivarc reviewed (approved):
Thank you!
/hold cancel 
Ping me if you decide to adjust the comment.

xpivarc commented on tests/migration.go:
```diff
@@ -72,6 +77,80 @@ func ConfirmVMIPostMigration(virtClient kubecli.KubevirtClient, vmi *v1.VirtualM
+		return true
+	}, 180*time.Second, 10*time.Second).Should(BeTrue(), "Some virt-handler pods survived the migration settings change")
+
+	return res
```
I think we need to check the new handlers are ready otherwise we will see this failure https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6036/pull-kubevirt-e2e-k8s-1.20-sig-network/1468980795975667712

xpivarc reviewed (commented):
/hold

stu-gott commented:
@xpivarc does this still need to have a hold label?

xpivarc commented on tests/migration.go:
```diff
@@ -72,6 +79,92 @@ func ConfirmVMIPostMigration(virtClient kubecli.KubevirtClient, vmi *v1.VirtualM
+		return true
+	}, 180*time.Second, 10*time.Second).Should(BeTrue(), "Some virt-handler pods survived the migration settings change")
+
+	// Ensure all virt-handlers are ready
```
We are only checking they are running not ready. It would be better to check for readiness but we don't advertise the device plugin readiness.

xpivarc reviewed (approved):
/hold cancel

xpivarc commented:
/retest

jean-edouard commented:
Moved tested virt-handler code from cmd to pkg and rebased

xpivarc reviewed (approved):
@jean-edouard Thank you!

jean-edouard commented:
/retest

rmohr commented:
Finally 😎😁
